### ISSUE TITLE
Questions on reproducing the reported results on MS COCO

### ISSUE THREAD

**Issue by shuijundadoudou** [Alibaba-MIIL/ASL#30]
Hi,

First, thank you for sharing the exciting work.

I was trying to reproduce the results on MS COCO dataset based on my own training framework. When I used cross entropy loss `loss_function=AsymmetricLoss(gamma_neg=0,** gamma_pos=0, clip=0)` to achieve the baseline. The result (with backbone of ResNet101) of mAP ~82.5% was achieved, which is quite similar to the result reported in Fig. 8 of the paper. 

Then, I replaced the loss function with `loss_function=AsymmetricLoss(gamma_neg=4, gamma_pos=1, clip=0.05)` -- all other hyper parameters were kept consistent. However, I only got the mAP result of ~82.1%. 

Also, the traditional focal loss `loss_function=AsymmetricLoss(gamma_neg=2, gamma_pos=2, clip=0)` can not outperform the baseline (~82.5%), given the same configurations. I am curious about the issue of my training process.

Could you also please share some training tricks? For example, a snippet of code on adjusting learning rate, training transforms similar to that used for validation [here](https://github.com/Alibaba-MIIL/ASL/blob/009ea6b2c70e8b935ea4a1877a0ce205644eab3b/validate.py#L57-L60), etc. Or, is there any suggestions?

Thank you.

**Comment by mrT23** [comment]
We honestly haven't encountered any case where ASL has not outperformed easily cross entropy.

some training tricks (most to them are quite standard), see if something resonant differently from your framework:
- for learning rate, we use one cycle policy (warmup + cosine decay) with Adam optimizer and max learning rate of ~2e-4 to 4e-4
- very important to use also EMA
- true weight decay of 1e-4 ("true" == no wd for batch norm and bias)
- we have our own augmentation package, but important to use at least standard AutoAugment.
- cutout of 0.5 (very important)
- squish resizing, not crop (important)
- try replacing resnet with TResNet. it will give you the same GPU speed, with higher accuracy

that's what I can think of at the top of my head.

**Comment by shuijundadoudou** [comment]
> We honestly haven't encountered any case where ASL has not outperformed easily cross entropy.
> 
> some training tricks (most to them are quite standard), see if something resonant differently from your framework:
> 
> * for learning rate, we use one cycle policy (warmup + cosine decay) with Adam optimizer and max learning rate of ~2e-4 to 4e-4
> * very important to use also EMA
> * true weight decay of 1e-4 ("true" == no wd for batch norm and bias)
> * we have our own augmentation package, but important to use at least standard AutoAugment.
> * cutout of 0.5 (very important)
> * squish resizing, not crop (important)
> * try replacing resnet with TResNet. it will give you the same GPU speed, with higher accuracy
> 
> that's what I can think of at the top of my head.

Thank you for the information. 

As I tried the cross entropy loss and the proposed ASL with same training configurations -- with only standard data augmentations, I suppose ASL should have produced at least better results. For these training tricks, I do believe they have potentials to improve the performance (both cross entropy and ASL). 

If I apply these tricks to ASL training, I also need to apply them to traditional cross entropy training to verify that the performance improvement comes from ASL (instead of these tricks). So, I am wondering if the experiments (for the same backbone) that produce the results in Fig. 8 used all these tricks (e.g. EMA, AutoAugment)? Also, from my understanding, both cross entropy and ASL in Fig. 8 are initialized by the corresponding models (ResNet/TresNet) trained on ImageNet, right?

Besides, could you please also specify some more details for the following hyper parameters, so that I do not have to try them all:
- how many steps are used for warmup?
- what is the size of cutout, the pixel length of the mask?
- for the true weight decay, are the BNs updated during training? Or, they are initialized with pretrained models (ResNet/TresNet on ImageNet) and fixed during MS-COCO training?

Thank you!

**Comment by mrT23** [comment]
I think that our approaches ("philosophy" :) ) for deep learning is a bit different

"training tricks" is a bit underwhelming name for the most important thing in deep learning. they are not extra methods your should choose whether to use or not. they are the essence, the bread-and-butter. i would be very proud if someone categorizes ASL as a good training trick.

training without proper augmentations, for example, is unacceptable in my opinion. the training would quickly overfit, no matter what your loss function is. EMA is also essential in any modern deep learning scheme. it usually outscores the regular model by ~1%, and generalize better.
when your baseline training tricks are not well calibrated, any additional trick, even if effective, will not give the full impact. without augmentations to prevent overfit, no loss function will save you.

my answers to your questions:
- always use proper augmentations, AutoAugment and above. always.
- always use EMA in addition to regular models. always
- warmup is usually 10-30% of the training.
- cutout is by far the most important augmentation in classification problems. without it all, my scores drop by 1%. i use a large cutout, with length half of the original picture
- BN are still updated and behave the same way with true wd. you just don't do l2 regularization on them. recommended wd: 1e-4

all of the training tricks i mentioned are used in this repository:
https://github.com/rwightman/pytorch-image-models
and i recommend reviewing it.

all the best
Tal

**Comment by SonDaoDuy** [comment]
Hello @shuijundadoudou, can you share the training setup to get ~ 82.5 mAP for the cross-entropy loss? for example, the optimizer, learning rate augmentation, and any training tricks?

**Comment by wuhy08** [comment]
> We honestly haven't encountered any case where ASL has not outperformed easily cross entropy.
> 
> some training tricks (most to them are quite standard), see if something resonant differently from your framework:
> 
> * for learning rate, we use one cycle policy (warmup + cosine decay) with Adam optimizer and max learning rate of ~2e-4 to 4e-4
> * very important to use also EMA
> * true weight decay of 1e-4 ("true" == no wd for batch norm and bias)
> * we have our own augmentation package, but important to use at least standard AutoAugment.
> * cutout of 0.5 (very important)
> * squish resizing, not crop (important)
> * try replacing resnet with TResNet. it will give you the same GPU speed, with higher accuracy
> 
> that's what I can think of at the top of my head.

Hi @mrT23 , thank you for the details.

When you say "squish resizing", do you mean that unlock the aspect ratio during resizing such that objects are stretched? 

What is the reason behind that? Is that because we could potentially crop some objects out of the image?

Since COCO dataset has positional info (bounding boxes, masks), does it make sense to use such info to mark certain objects as negative if those objects are cropped out?

**Comment by mrT23** [comment]
read more about squish vs crop resizing here:
![image](https://user-images.githubusercontent.com/21198860/103544786-49d3f800-4ea9-11eb-99f1-f6c1ea964d6d.png)
https://forums.fast.ai/t/resize-instead-of-crop/28680/10
https://docs.fast.ai/vision.augment.html#Resize-with-crop,-pad-or-squish

crop-resizing works basically only on ImageNet, because the vast majority of objects there are zoomed-in and in the center.
its a common mistake to automatically use ImageNet resizing to other datasets.
however,  squish-resizing usually gives better results, and is less prone for catastrophic mistakes

**Comment by wuhy08** [comment]
Thank you @mrT23 for the elaboration. Really helpful.

**Comment by adam-dziedzic** [comment]
> Hello @shuijundadoudou, can you share the training setup to get ~ 82.5 mAP for the cross-entropy loss? for example, the optimizer, learning rate augmentation, and any training tricks?

This would be helpful.

**Comment by mrT23** [comment]
I agree. 

We cannot share our code as-is due to commercial limitations, but once a public code will be shared, we can try to help improve it and get results similar to the ones in the article

**Comment by GhostWnd** [comment]
I also want to reproduce the result on MSCOCO, but due to my limitation on GPU resourses and limited time, I resize the img to 224*224.  I use no training tricks and set the learning rate 1e-4 constantly. At first I thought even if I can't get mAP 86.6, I can get at least mAP 70, which is enough for me. But after about 100 iteration in Epoch 1, the loss decerases from 230 to 90 and doesn't decrease anymore, the mAP on validation is 6, very low. I wonder whether it's because I don't implement the tricks or I should not resize the images to 224 or simply the code is wrong.

Here are my training code, and I don't change other files like tresnet.py or so, I will be very grateful if you point out my problems, it has puzzled me for quite a long time.

# -*- coding: utf-8 -*-
parser = argparse.ArgumentParser(description='PyTorch ImageNet Training')
parser.add_argument('data', metavar='DIR', help='path to dataset')
parser.add_argument('--model-name', default='tresnet_l')
parser.add_argument('--model-path', default='./TRresNet_L_448_86.6.pth', type=str)
parser.add_argument('--num-classes', default=80)
parser.add_argument('-j', '--workers', default=4, type=int, metavar='N',
                    help='number of data loading workers (default: 16)')
parser.add_argument('--image-size', default=224, type=int,
                    metavar='N', help='input image size (default: 448)')
parser.add_argument('--thre', default=0.8, type=float,
                    metavar='N', help='threshold value')
parser.add_argument('-b', '--batch-size', default=32, type=int,
                    metavar='N', help='mini-batch size (default: 16)')
parser.add_argument('--print-freq', '-p', default=64, type=int,
                    metavar='N', help='print frequency (default: 64)')

def main():
    args = parser.parse_args()
    args.batch_size = args.batch_size

    # setup model
    print('creating model...')
    args.do_bottleneck_head = False
    model = create_model(args).cuda()
    model.train()
    print('done\n')

    # Data loading code
    normalize = transforms.Normalize(mean=[0, 0, 0],
                                     std=[1, 1, 1])

    instances_path_val = os.path.join(args.data, 'annotations/instances_val2017.json')
    instances_path_train = os.path.join(args.data, 'annotations/instances_train2017.json')
    
    data_path_val = os.path.join(args.data, 'val2017')
    data_path_train = os.path.join(args.data, 'train2017')
    val_dataset = CocoDetection(data_path_val,
                                instances_path_val,
                                transforms.Compose([
                                    transforms.Resize((args.image_size, args.image_size)),
                                    transforms.ToTensor(),
                                    normalize,
                                ]))
    train_dataset = CocoDetection(data_path_train,
                                instances_path_train,
                                transforms.Compose([
                                    transforms.Resize((args.image_size, args.image_size)),
                                    transforms.ToTensor(),
                                    normalize,
                                ]))

    print("len(val_dataset)): ", len(val_dataset))
    print("len(train_dataset)): ", len(train_dataset))
    val_loader = torch.utils.data.DataLoader(
        val_dataset, batch_size=args.batch_size, shuffle=False,
        num_workers=args.workers, pin_memory=False)
    train_loader = torch.utils.data.DataLoader(
        train_dataset, batch_size=args.batch_size, shuffle=True,
        num_workers=args.workers, pin_memory=False)
    
    criterion = AsymmetricLoss()
    params = model.parameters()
    optimizer = torch.optim.Adam(params, lr=0.0001)
    total_step = len(train_loader)
    
    highest_mAP = 0
    trainInfoList = []
    Sig = torch.nn.Sigmoid()
    for epoch in range(5):
        for i, (inputData, target) in enumerate(train_loader):
            model.train()
            inputData = inputData.cuda()
            target = target.cuda()
            target = target.max(dim=1)[0]
            #output = torch.nn.Sigmoid(model(inputData))
            output = Sig(model(inputData))
            loss = criterion(output, target)
            
            model.zero_grad()
            loss.backward()
            optimizer.step()
            
            #store information
            if i % 10 == 0:
                trainInfoList.append([epoch, i, loss.item()])
                print('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}'
                      .format(epoch, 5, i, total_step, loss.item()))
                
            if (i+1) % 100 == 0:
                #store model
                torch.save(model.state_dict(), os.path.join(
                    'models/', 'model-{}-{}.ckpt'.format(epoch+1, i+1)))
                mAP_score = validate_multi(val_loader,  model, args)
                if mAP_score > highest_mAP:
                    highest_mAP = mAP_score
                    print('current highest_mAP = ', highest_mAP)
                    torch.save(model.state_dict(), os.path.join(
                            'models/', 'model-highest.ckpt')) 
    
def validate_multi(val_loader, model, args):
    print("starting actuall validation")
    batch_time = AverageMeter()
    prec = AverageMeter()
    rec = AverageMeter()
    mAP_meter = AverageMeter()

    Sig = torch.nn.Sigmoid()

    end = time.time()
    tp, fp, fn, tn, count = 0, 0, 0, 0, 0
    preds = []
    targets = []
    model.eval()
    for i, (input, target) in enumerate(val_loader):
        target = target
        target = target.max(dim=1)[0]
        # compute output
        with torch.no_grad():
            output = Sig(model(input.cuda())).cpu()

        # for mAP calculation
        preds.append(output.cpu())
        targets.append(target.cpu())

        # measure accuracy and record loss
        pred = output.data.gt(args.thre).long()

        tp += (pred + target).eq(2).sum(dim=0)
        fp += (pred - target).eq(1).sum(dim=0)
        fn += (pred - target).eq(-1).sum(dim=0)
        tn += (pred + target).eq(0).sum(dim=0)
        count += input.size(0)

        this_tp = (pred + target).eq(2).sum()
        this_fp = (pred - target).eq(1).sum()
        this_fn = (pred - target).eq(-1).sum()
        this_tn = (pred + target).eq(0).sum()

        this_prec = this_tp.float() / (
            this_tp + this_fp).float() * 100.0 if this_tp + this_fp != 0 else 0.0
        this_rec = this_tp.float() / (
            this_tp + this_fn).float() * 100.0 if this_tp + this_fn != 0 else 0.0

        prec.update(float(this_prec), input.size(0))
        rec.update(float(this_rec), input.size(0))

        # measure elapsed time
        batch_time.update(time.time() - end)
        end = time.time()

        p_c = [float(tp[i].float() / (tp[i] + fp[i]).float()) * 100.0 if tp[
                                                                             i] > 0 else 0.0
               for i in range(len(tp))]
        r_c = [float(tp[i].float() / (tp[i] + fn[i]).float()) * 100.0 if tp[
                                                                             i] > 0 else 0.0
               for i in range(len(tp))]
        f_c = [2 * p_c[i] * r_c[i] / (p_c[i] + r_c[i]) if tp[i] > 0 else 0.0 for
               i in range(len(tp))]

        mean_p_c = sum(p_c) / len(p_c)
        mean_r_c = sum(r_c) / len(r_c)
        mean_f_c = sum(f_c) / len(f_c)

        p_o = tp.sum().float() / (tp + fp).sum().float() * 100.0
        r_o = tp.sum().float() / (tp + fn).sum().float() * 100.0
        f_o = 2 * p_o * r_o / (p_o + r_o)

        if i % args.print_freq == 0:
            print('Test: [{0}/{1}]\t'
                  'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                  'Precision {prec.val:.2f} ({prec.avg:.2f})\t'
                  'Recall {rec.val:.2f} ({rec.avg:.2f})'.format(
                i, len(val_loader), batch_time=batch_time,
                prec=prec, rec=rec))
            print(
                'P_C {:.2f} R_C {:.2f} F_C {:.2f} P_O {:.2f} R_O {:.2f} F_O {:.2f}'
                    .format(mean_p_c, mean_r_c, mean_f_c, p_o, r_o, f_o))

    print(
        '--------------------------------------------------------------------')
    print(' * P_C {:.2f} R_C {:.2f} F_C {:.2f} P_O {:.2f} R_O {:.2f} F_O {:.2f}'
          .format(mean_p_c, mean_r_c, mean_f_c, p_o, r_o, f_o))

    mAP_score = mAP(torch.cat(targets).numpy(), torch.cat(preds).numpy())
    print("mAP score:", mAP_score)

    return mAP_score
if __name__ == '__main__':
    main()

**Comment by 1006927966** [comment]
@GhostWnd   hello你的问题发现什么原因了吗 我也出现了这个问题  谢谢

**Comment by ChenAnno** [comment]
> We honestly haven't encountered any case where ASL has not outperformed easily cross entropy.
> 
> here are some training tricks we used (they are quite standard and can be found also in public repositories like [this](https://github.com/rwightman/pytorch-image-models)), see if something resonant differently from your framework:
> 
> * for learning rate, we use one cycle policy (warmup + cosine decay) with Adam optimizer and max learning rate of ~2e-4 to 4e-4
> * very important to use also EMA
> * true weight decay of 1e-4 ("true" == no wd for batch norm and bias)
> * we have our own augmentation package, but important to use at least standard AutoAugment.
> * cutout of 0.5 (very important)
> * squish resizing, not crop (important)
> * try replacing resnet with TResNet. it will give you the same GPU speed, with higher accuracy
> 
> that's what I can think of at the top of my head.

Great thanks for your details, which really helps me a lot!
